sys/usb/usbus/usbus.c

```diff
@@ -0,0 +1,667 @@
+            break;
+        case USB_SETUP_REQUEST_RECIPIENT_INTERFACE:
+            recv_interface_setup(usbus, ep, pkt);
+            break;
+        default:
+            DEBUG("usbus: Unhandled setup request\n");
+    }
+}
+
+static void *_usbus_thread(void *args)
+{
+    usbus_t *usbus = (usbus_t*)args;
+
+    mutex_lock(&usbus->lock);
+    usbdev_t *dev = usbus->dev;
+    usbus->pid = sched_active_pid;
+    usbus->addr = 0;
+    usbus->strings = NULL;
+    usbus->iface = NULL;
+    usbus->str_idx = 1;
+    usbus->setup_state = USBUS_SETUPRQ_READY;
+    mutex_init(&usbus->lock);
+    msg_t msg, msg_queue[_USBUS_MSG_QUEUE_SIZE];
+    DEBUG("usbus: starting thread %i\n", sched_active_pid);
+    /* setup the link-layer's message queue */
+    msg_init_queue(msg_queue, _USBUS_MSG_QUEUE_SIZE);
+    /* register the event callback with the device driver */
+    dev->cb = _event_cb;
+    /* initialize low-level driver */
+    dev->driver->init(dev);
+    dev->context = usbus;
+
+    usbus->in = usbus->dev->driver->new_ep(usbus->dev, USB_EP_TYPE_CONTROL, USB_EP_DIR_IN, USBUS_MAX_SIZE);
+    usbus->out = usbus->dev->driver->new_ep(usbus->dev, USB_EP_TYPE_CONTROL, USB_EP_DIR_OUT, USBUS_MAX_SIZE);
+    usbus->in->cb = _event_ep0_cb;
+    usbus->out->cb = _event_ep0_cb;
+    usbus->in->context = usbus;
+    usbus->out->context = usbus;
+
+    usbus->in->driver->init(usbus->in);
+    usbus->out->driver->init(usbus->out);
+    _usbus_config_ep0(usbus);
+    usbus_add_string_descriptor(usbus, &usbus->config, USB_CONFIG_CONFIGURATION_STR);
+    usbus_add_string_descriptor(usbus, &usbus->product, USB_CONFIG_PRODUCT_STR);
+    usbus_add_string_descriptor(usbus, &usbus->manuf, USB_CONFIG_MANUF_STR);
+
+    usbus->state = USBUS_STATE_DISCONNECT;
+    mutex_unlock(&usbus->lock);
+
+    /* TODO: signal ready from application */
+    xtimer_sleep(1);
+    usbopt_enable_t enable = USBOPT_ENABLE;
+    dev->driver->set(dev, USBOPT_ATTACH, &enable, sizeof(usbopt_enable_t));
+
+    while (1) {
+        msg_receive(&msg);
+        /* dispatch netdev, MAC and gnrc_netapi messages */
+        switch (msg.type) {
+            case USBUS_MSG_TYPE_EVENT:
+                dev->driver->esr(dev);
+                break;
+            case USBUS_MSG_TYPE_EP_EVENT:
+                {
+                    usbdev_ep_t *ep = (usbdev_ep_t*)msg.content.ptr;
+                    ep->driver->esr(ep);
+                }
+              break;
+            default:
+                DEBUG("usbus: unhandled event\n");
+                break;
+        }
+    }
+    return NULL;
+}
+
+/* USB event callback */
+void _event_cb(usbdev_t *usbdev, usbdev_event_t event)
+{
+    usbus_t *usbus = (usbus_t *)usbdev->context;
+    if (event == USBDEV_EVENT_ESR) {
+        msg_t msg = { .type = USBUS_MSG_TYPE_EVENT,
+                      .content = { .ptr = usbdev } };
+
+        if (msg_send(&msg, usbus->pid) <= 0) {
+            puts("usbus: possibly lost interrupt.");
```

Comment:
I'd seriously consider using thread flags for this. You'd disable this failure case entirely, thread flags cannot get lost.

An arriving message also sets a thread flag, so a thread can wait for both:

```
thread_flag_t flags = thread_flag_wait_any(0xffff);
if (flags & THREAD_FLAG_MSG_WAITING) {
  msg_t msg
  while (msg_try_receive(&msg) == 1) {
     // handle netapi messages
  }
}
  if (flags & YOUR_FLAG_SET_BY_ISR) {
    /// handle
  }
}
```

Reply:
I understand if you'd rather not go into this for now, can be changed later.

Reply:
This is actually one of the issues where I'd really would like to have some feedback on. I've used messages here because it's what I'm most familiar with from gnrc, but using thread flags or events are both open to suggestions for me.

Reply:
After some out-of-band discussion it became clear that it should be easy to move to thread flags with a flag for the generic usbdev ISR and a flag for endpoint specific ISR's. A uint32_t can keep track of the endpoints that require servicing (USB has 32 endpoints max, 16 IN and 16 OUT). This has as a consequence that usbus needs fast access to the usbdev endpoints, preferably O(1) complexity to resolve the thread flag back to an endpoint.

One solution here is to allocate the `usbus_endpoint_t` (not to be confused with `usbdev_ep_t`) as a static array. This could then reuse the `USBDEV_NUM_ENDPOINTS`, matching the number of endpoints allocated by the peripheral.

Reply:
@kaspar030 I rewrote the loop to use thread flags. 2 flags are used as described above. Additionally I've added handling for `events` to allow signalling events to usb functionality from outside or inside the usbus thread.